src/redprl/refiner.fun

```diff
@@ -1015,8 +1015,224 @@ struct
+            val J = H >> CJ.EQ ((tube0,tube1), ty)
+          in
+            List.map (fn j => #1 (makeGoal (([(w, P.DIM)], []) |> j)))
+                     (Restriction.One J ext eps)
+          end
+
+        val group0 = groupTubes exts0 tubes0
+        val group1 = groupTubes exts1 tubes1
+
+        val interTubeGoals =
+          listToTel
+            (ListMonad.bind
+               (fn ((ext, eps, tube0), (_, _, tube1)) => interTube ext eps tube0 tube1)
+               (ListPair.zipEq (group0, group1)))
+      in
+        T.append (T.empty >: goalTy >: goalCap)
+                 (T.append interTubeGoals (T.append (intraTubeGoals H w ty group0)
+                                                    (tubeCapGoals H ty r0 cap0 group0)))
+        #> trivial
+      end
+
+    fun CapEq alpha jdg =
+      let
+        val _ = RedPrlLog.trace "HCom.CapEq"
+        val H >> CJ.EQ ((lhs, rhs), ty) = jdg
+        val Syn.HCOM (exts, (r, r'), ty0, cap, tubes) = Syn.out lhs
+        val () = assertParamEq "HCom.CapEq source and target of direction" (r, r')
+        val () = assertAlphaEq (ty0, ty)
+        val group = groupTubes exts tubes
+
+        val (goalTy, _) = makeGoal @@ H >> CJ.TYPE ty
+        val (goalCap, _) = makeGoal @@ H >> CJ.MEM (cap, ty)
+        val (goalEq, _) = makeGoal @@ H >> CJ.EQ ((cap, rhs), ty)
+
+        val w = alpha 0
+      in
+        T.append (T.empty >: goalTy >: goalCap >: goalEq)
+                 (T.append (intraTubeGoals H w ty group)
+                           (tubeCapGoals H ty r cap group))
+        #> trivial
+      end
+
+    fun TubeEq i alpha jdg =
+      let
+        val _ = RedPrlLog.trace "HCom.TubeEq"
+        val H >> CJ.EQ ((lhs, rhs), ty) = jdg
+        val Syn.HCOM (exts, (r, r'), ty0, cap, tubes) = Syn.out lhs
+        val () = assertAlphaEq (ty0, ty)
+        val j =
+          case List.nth (exts, i) of
+             P.APP P.DIM0 => 0
+           | P.APP P.DIM1 => 1
+           | _ => raise Fail "HCom.TubeEq extent"
+
+        val (u, tube) = List.nth (tubes, 2 * i + j)
+
+        val group = groupTubes exts tubes
+
+        val (goalTy, _) = makeGoal @@ H >> CJ.TYPE ty
+        val (goalCap, _) = makeGoal @@ H >> CJ.MEM (cap, ty)
+        val (goalEq, _) = makeGoal @@ H >> CJ.EQ ((substSymbol (r', u) tube, rhs), ty)
+
+        val w = alpha 0
+      in
+        T.append (T.empty >: goalTy >: goalCap >: goalEq)
+                 (T.append (intraTubeGoals H w ty group)
+                           (tubeCapGoals H ty r cap group))
+        #> trivial
+      end
+
+    (* Return index of first element in the list satisfying p. *)
+    fun indexOf p =
+      let
+        fun go acc [] = raise List.Empty
+          | go acc (x :: l) =
+            if p x then acc
+            else go (acc + 1) l
+      in
+        go 0
+      end
+
+    (* Search for an index of a constant extent in an hcom, to pass to HCom.TubeEq. *)
+    fun FindTubeEq alpha jdg =
```

Comment:
Good idea, factoring this out and having the primitive rule take `i` as a parameter.